source/opt/ccp_pass.cpp

```diff
@@ -120,58 +120,37 @@ SSAPropagator::PropStatus CCPPass::VisitAssignment(ir::Instruction* instr) {
     return MarkInstructionVarying(instr);
   }
 
-  // Otherwise, see if the RHS of the assignment folds into a constant value.
-  std::vector<uint32_t> cst_val_ids;
-  bool missing_constants = false;
-  bool varying_values = false;
-  instr->ForEachInId([this, &cst_val_ids, &missing_constants,
-                      &varying_values](uint32_t* op_id) {
+  // See if the RHS of the assignment folds into a constant value.
+  auto map_func = [this](uint32_t id) {
+    auto it = values_.find(id);
+    if (it == values_.end() || it->second == kVaryingSSAId) {
+      return id;
+    }
+    return it->second;
+  };
+  ir::Instruction* folded_inst = opt::FoldInstruction(instr, map_func);
+  if (folded_inst != nullptr && folded_inst->IsConstant()) {
```

Comment:
What if we improve the folder to generate non-constant instructions. Then we would leak the folded_inst in those cases.
For example, what if the folder implements a form of strength reduction.

Reply:
I had thought of this a little.  The problem is assessing who is the owner of the instruction pointer to by `folded_inst`.  It cannot be `folded_inst` because most of the time we will be returning a value that already exists in the code, like an OpConstant.  Even though I did not document it, if `FoldInstruction` creates a new instruction, it will have to link it into the code somewhere so that it is owned by a basic block or other list.  I did not document that because I am still unsure how to decide which block it belongs to, should there be a requirement that the input instruction be in a list, and the new instructions be placed in that list?

I could say that for now, `FoldInstruction` will not create a new instruction, but that will probably change soon enough.

Thanks for bring this up, because I forgot to.  It should be discussed.

Reply:
I misspoke.  Current if a new instruction is created it is a constant, and it is added to the section that holds all of the constants.  When we create other instructions, it will become more complicated.

Reply:
This is correct but fragile.   I'd recommend being a little more defensive:

```
if (folded_inst) {
  if (folded_inst->IsConstant()) { ... }
  else { 
    assert and a TODO.
  }
}
```
But I can put that in in a subsequent CL.

Reply:
I've updated the folding routines, and created two version.  One that will only return constants, which is called from here.  I've added an assert to make sure the folded instruction is a constant.